crates/sui-core/src/state_accumulator.rs

```diff
@@ -0,0 +1,588 @@
+            .watermarks
+            .get(&CheckpointWatermark::HighestReadyToHash)?
+        {
+            Ok(Some(highest_ready))
+        } else {
+            Ok(None)
+        }
+    }
+
+    pub fn update_highest_hashed_checkpoint_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestHashed, checkpoint_seq_num)
+    }
+
+    pub fn update_highest_accumulated_checkpoint_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestAccumulated, checkpoint_seq_num)
+    }
+
+    pub fn update_highest_checkpoint_ready_to_hash_seq_number(
+        &self,
+        checkpoint_seq_num: &CheckpointSequenceNumber,
+    ) -> Result<(), TypedStoreError> {
+        self.watermarks
+            .insert(&CheckpointWatermark::HighestReadyToHash, checkpoint_seq_num)
+    }
+}
+
+/// Thread safe clone-able handle for enqueueing items to Accumulator
+#[derive(Clone)]
+pub struct EnqueueHandle {
+    sender: mpsc::Sender<State>,
+    store: Arc<StateAccumulatorStore>,
+}
+
+impl EnqueueHandle {
+    /// Enqueue checkpoint state for accumulation. A checkpoint shall
+    /// be processed (accumulated both individually and towards
+    /// the root state hash) exactly once
+    ///
+    /// NOTE: When called from executor, this should be called before we increment
+    /// highest executed checkpoint watermark, otherwise we may inefficiently read effects
+    /// from the effects store rather than using effects passed in memory throuhg enqueue.
+    /// This may happen due to a race between scheduling new checkpoints for hashing and
+    /// checking the queue.
+    pub async fn enqueue(&self, state: State) -> SuiResult {
+        let seq_num = state.checkpoint_seq_num;
+
+        let highest = self.store.get_highest_checkpoint_ready_to_hash_seq_number().unwrap();
+
+        match highest {
+            // Check to ensure items are enqueued in order and no checkpoint sequence
+            // numbers are skipped
+            Some(highest) => {
+                match seq_num - highest {
+                    // next checkpoint in sequence - happy path
+                    1 => {},
+                    
+                    u64::MIN..=0 => {
+                        warn!("Checkpoint effects for sequence number {seq_num:?} already enqueued for accumulation");
+                        return Ok(());
+                    },
+                    
+                    2..=u64::MAX => panic!("Checkpoint seq num {seq_num:?} contents enqueued for accumulation out of order"),
+                }
+            }
+            None => {
+                if seq_num != 0 {
+                    panic!("Checkpoint seq num {seq_num:?} contents enqueued for accumulation out of order");
```

Comment:
Do we need to use an approach that is so reliant on panics? Checkpoints can be hashed in any order, only accumulation requires consecutive checkpoints.

I don't feel too strongly about this but I do wonder if the code would be more straightforward if we just hashed anything that wasn't already hashed, and then later did the accumulation as it becomes possible.

Reply:
Fixed.